public/app/features/variables/editor/reducer.ts

```diff
@@ -22,54 +30,74 @@ const variableEditorReducerSlice = createSlice({
       state.id = '';
     },
-    variableEditorMounted: (state: VariableEditorState, action: PayloadAction<{ name: string }>) => {
+    variableEditorMounted: (
+      state: VariableEditorState<QueryVariableEditorState>,
+      action: PayloadAction<{ name: string }>
+    ) => {
       state.name = action.payload.name;
     },
-    variableEditorUnMounted: (state: VariableEditorState, action: PayloadAction<VariablePayload>) => {
+    variableEditorUnMounted: (
+      state: VariableEditorState<QueryVariableEditorState>,
+      action: PayloadAction<VariablePayload>
+    ) => {
       return initialVariableEditorState;
     },
     changeVariableNameSucceeded: (
-      state: VariableEditorState,
+      state: VariableEditorState<QueryVariableEditorState>,
       action: PayloadAction<VariablePayload<{ newName: string }>>
     ) => {
       state.name = action.payload.data.newName;
       delete state.errors['name'];
       state.isValid = Object.keys(state.errors).length === 0;
     },
     changeVariableNameFailed: (
-      state: VariableEditorState,
+      state: VariableEditorState<QueryVariableEditorState>,
       action: PayloadAction<{ newName: string; errorText: string }>
     ) => {
       state.name = action.payload.newName;
       state.errors.name = action.payload.errorText;
       state.isValid = Object.keys(state.errors).length === 0;
     },
     addVariableEditorError: (
-      state: VariableEditorState,
+      state: VariableEditorState<QueryVariableEditorState>,
       action: PayloadAction<{ errorProp: string; errorText: any }>
     ) => {
       state.errors[action.payload.errorProp] = action.payload.errorText;
       state.isValid = Object.keys(state.errors).length === 0;
     },
-    removeVariableEditorError: (state: VariableEditorState, action: PayloadAction<{ errorProp: string }>) => {
+    removeVariableEditorError: (
+      state: VariableEditorState<QueryVariableEditorState>,
+      action: PayloadAction<{ errorProp: string }>
+    ) => {
       delete state.errors[action.payload.errorProp];
       state.isValid = Object.keys(state.errors).length === 0;
     },
     changeVariableEditorExtended: (
-      state: VariableEditorState,
+      state: VariableEditorState<QueryVariableEditorState>,
       action: PayloadAction<{ propName: string; propValue: any }>
     ) => {
       state.extended = {
         ...state.extended,
         [action.payload.propName]: action.payload.propValue,
       };
     },
+    updateQueryVariableDatasource: (
+      state: VariableEditorState<QueryVariableEditorState>,
+      action: PayloadAction<{ datasource: DataSourceApi; queryEditor: VariableQueryEditorType; variable: unknown }>
+    ) => {
+      if (!state.extended) {
+        state.extended = {};
+      }
+
+      state.extended.dataSource = action.payload.datasource;
+      state.extended.VariableQueryEditor = action.payload.queryEditor;
+    },
```

Comment:
What this is doing is taking a general reducer that works for all variable types and changes it to a more specific reducer for QueryVariables. I don't think this is a change we want to introduce.

Reply:
Thanks for the look - makes sense! That was the piece I was struggling to see, how this all worked with the various types of variables. A thing that made this a bit trickier is the _very_ generous types of these actions (and state) have that made it tricky to know what things should be and how it's used. 

What's your sense of how possible a change like this would be, to dispatch both the editor and datasource in the same action? From what you're saying right now it seems like the only/most straight forward approach would be to continue the pattern of the untyped `changeVariableEditorExtended` action, possibly with one to set multiple properties at once? What's your thoughts on that?

Reply:
I guess you want to have one action because you're worried about performance? Not sure that is a real problem here. But I would continue down the path of the "untyped" action. The other change would be to introduce specific variable type editor reducers and although that would be awesome I'm not sure it's worth it.

Reply:
Performance is a "nice to have" outcome that I'm kind of thinking about, but the main thing I'm thinking about through all this is "what are the individual atomic changes to the state that are required as a result of calling this action" (phew!). When changing the datasource from Prometheus to Azure Monitor does it ever make sense to dispatch once action (the datasource) without the other (the query editor)?

I think the original bug report, of the datasource being out of sync from the query editor is directly from these two action (which we consider to be the "same change", I think?) being split over two redux dispatches, and two renders.

Reply:
I don't think having seperate reducers for different types is the right direction to go down (seems like a lot of work!!), but I'm wondering if something like a "tagged union" could be helpful here?

Granted I still don't fully grasp the big picture of how exactly the different types of variable state all work together, but I'm wondering if (hypothetically) something like this could work:

```typescript
interface QueryVariableExtendedState {
  type: "Query";
  datasource: DataSourceApi;
  VariableQueryEditor?: VariableQueryEditorType;
}

interface AdHocVariableExtendedState {
  type: "AdHoc";
  ...
}

type VariableEditorExtendedState = QueryVariableExtendedState | AdHocVariableExtendedState;

interface VariableEditorState  {
  ...
  extended: VariableEditorExtendedState;
}
```

You can still have generic actions that dispatch `VariableEditorExtendedState` for any type, and they can change up to be logically complete changes, rather than one prop at a time.

Reply:
This last suggestion with the `tagged union` seems like that could work. Give it a go and let's see where it takes us.